### DISCUSSION TITLE
Multiple collections on one page

### DISCUSSION THREAD

**Discussion by djdrysdale** [11ty/eleventy#2114] [Q&A]
I'm building my site in Eleventy and can get the basics fine but I am having trouble with collections. I want to display the three most recent posts from two different parts of my site on the home page. I have created pages that use one collection (say, posts) but I can't figure out how to do two. Is this possible?

**Comment by pdehaan** [comment]
Not sure I understand what you mean by "from two different parts of my site"...
Are you trying to get the first 3 items from the `collections.posts` collection?
Or are you trying to merge two different collections and get the first 3 items?

**Reply by djdrysdale** [reply]
On [my site](http://www.mobydiction.ca) I publish two different kinds of content, Posts and Notes. These are sourced from two separate folders of markdown content. There are some distinctions between the two--for instance, Posts will have featured images and notes are only text. I have two pages on my site that collect these individually ([Posts](https://mobydiction.ca/posts/), [Notes](https://mobydiction.ca/notes/)) but I would like to be able to display the three most recent items from each of those types of content on the home page. 

I'm already doing it on the [homepage](https://mobydiction.ca/) by using pagination on collections.posts, but I can't figure out how to add content from collection.notes as well.  

So yes, I want to get the first three items from collections.posts, but I also want to get the first three items from collections.notes.

**Reply by pdehaan** [reply]
By chance, is your site https://github.com/djdrysdale/digital-garden?

I managed to get that to show some recent posts+notes on the homepage with a custom filter and a couple minor tweaks.

### /.eleventy.js
Here's a little filter, `head`, that returns a subset of an array. So if you want the first five items in an array/collection we can use `head: 5`, or if you want the last 3 items, we can use `head: -3` (we'll use this in /layouts/index.html below)
```js
    eleventyConfig.addFilter("head", function (arr=[], count=5) {
        if (count < 0) {
            return arr.slice(count);
        }
        return arr.slice(0, count);
    });
```

### /layouts/index.html
First we create two new variables, `recent_posts` and `recent_notes`, which take the last 3 items from their respective collections. And then we pass those variables into our include file as the `items` variable. You seem to be using .html and .md templates, and currently (Eleventy v0.12.1) uses LiquidJS as the engine for those. So see https://liquidjs.com/tags/include.html#Passing-Variables. If you switch to Nunjucks, you'd probably need to revisit this code (or switch to macros or something). We could turn this into a custom shortcode instead of an include, but then you'd need to do the `recentposts.html` templating in JavaScript instead of LiquidJS/Nunjucks. 🤷 
```liquid
                 <div class="row py-4">
                   <h3>Recent Posts</h3>
                   {% assign recent_posts = collections.posts | head: -3 %}
                   {% include recentposts.html, items: recent_posts %}
                 </div>
                 <div class="row py-4">
                   <h3>Recent Notes</h3>
                   {% assign recent_notes = collections.notes | head: -3 %}
                   {% include recentposts.html, items: recent_notes %}
                 </div>
```

### /includes/recentposts.html
Finally, here's the updated ./includes/recentposts.html file. Instead of looping over a globally scoped `pagination.items`, we loop over our supplied `items` array/collection. I also changed `note.data.description` and `note.url` references to `post.data.description` and `post.url` since there wasn't a `note` variable so those didn't seem to be rendering correctly.
```liquid
  {% for post in items %}
    <div class="card-body col-12 col-md-4 border-bottom py-10 px-10">
        <p class="card-text">
          <a href="{{ post.url }}">{{ post.data.title }}</a>
        </p>
        <p class="card-text">
          {{ post.data.description }}
        </p>
        <p>
          <a href="{{ post.url }}">Read more...</a>
        </p>
    </div>
  {% endfor %}
```

Oh, and I guess you should also remove that pagination code in /index.md. Which is nice because now it writes 412 files instead of 424 files, since that pagination was creating a bunch of files named _site/[1..12]/index.html because of the pagination.
> _Writing `_site/12/index.html` from ./index.md._

```liquid
---
layout: index
title: Moby Diction
---

Hi. I'm David Drysdale.

I'm a design strategist and researcher based in London, Ontario, Canada.

It's my mission to help people, teams, and organizations think more strategically, make better decisions, and achieve better results.
```

---

![Moby_Diction_•_Moby_Diction](https://user-images.githubusercontent.com/557895/142769961-563f3390-8643-40f3-a39c-25d9d883b5cc.png)

<details>
<summary>And here's the rich <kbd>git diff</kbd> if that's your thing</summary>

```diff
diff --git a/.eleventy.js b/.eleventy.js
index 6285daf..e94f818 100644
--- a/.eleventy.js
+++ b/.eleventy.js
@@ -32,6 +32,13 @@ module.exports = function(eleventyConfig) {
         return collection.getFilteredByGlob(["notes/**/*.md", "index.md"]);
     });
 
+    eleventyConfig.addFilter("head", function (arr=[], count=5) {
+        if (count < 0) {
+            return arr.slice(count);
+        }
+        return arr.slice(0, count);
+    });
+
     eleventyConfig.addPassthroughCopy('assets');
 
     return {
diff --git a/includes/recentposts.html b/includes/recentposts.html
index 861e02d..6e8cd72 100644
--- a/includes/recentposts.html
+++ b/includes/recentposts.html
@@ -1,14 +1,14 @@
 
-  {% for post in pagination.items %}
+  {% for post in items %}
     <div class="card-body col-12 col-md-4 border-bottom py-10 px-10">
         <p class="card-text">
           <a href="{{ post.url }}">{{ post.data.title }}</a>
         </p>
         <p class="card-text">
-          {{ note.data.description }}
+          {{ post.data.description }}
         </p>
         <p>
-          <a href="{{ note.url }}">Read more...</a>
+          <a href="{{ post.url }}">Read more...</a>
         </p>
 
     </div>
diff --git a/layouts/index.html b/layouts/index.html
index fe37659..4842547 100644
--- a/layouts/index.html
+++ b/layouts/index.html
@@ -49,7 +49,13 @@
                  </article>
                  <div class="row py-4">
                    <h3>Recent Posts</h3>
-                   {% include recentposts.html %}
+                   {%- assign recent_posts = collections.posts | head: -3 -%}
+                   {% include recentposts.html, items: recent_posts %}
+                 </div>
+                 <div class="row py-4">
+                   <h3>Recent Notes</h3>
+                   {%- assign recent_notes = collections.notes | head: -3 -%}
+                   {% include recentposts.html, items: recent_notes %}
                  </div>
                </div>
```
</details>

**Reply by pdehaan** [reply]
So, a few random observations while I stare at that glorious screenshot.
1. The order of the items seems to be reversed from your current site. We might need to add a `reverse` filter in there somewhere.
    ```liquid
    {% assign recent_posts = collections.posts | head: -3 | reverse %}
    {% include recentposts.html, items: recent_posts %}
    ```
1. Not sure why "Moby Diction" is a note. But from the looks of your .eleventy.js file, the Notes collection is a custom collection which specifically includes the root index.md:
    ```js
    return collection.getFilteredByGlob(["notes/**/*.md", "index.md"]);
    ```
    I was able to add `tags: ["notes"],` into your notes/notes.11tydata.js file to have it auto tag every entry in the "notes" collection (and could therefore delete the custom collection from the .eleventy.js file), but then that caused a build error w/ ./notes/index.md having a circular reference (`Dependency Cycle Found: ___TAG___all -> ./notes/index.md -> ___TAG___notes -> ./notes/index.md`), so I had to exclude that index template from the collection via `eleventyExcludeFromCollections: true` frontmatter.
1. Looking at the description for the "Summary: The Opposable Mind by Roger L. Martin" post, it looks like the excerpt might have some markdown in it. I tweaked that in the /includes/recentposts.html file using:
    ```liquid
    {{ post.data.description | markdownify }}
    ```
    But that was throwing build errors until I removed the "/index.md" from the Notes collection since it doesn't have a description, and Markdownify doesn't seem to know how to markdownify `null` or `undefined`.

**Reply by pdehaan** [reply]
And I also noticed some URLs aren't being slugified (specifically it looks like the /notes/ directory files):

http://localhost:8080/notes/The%20heroine's%20journey%20narrates%20the%20establishment%20of%20community%20and%20connection/

---

Or a really interesting/subtle issue is your bio in your [index.md homepage](https://github.com/djdrysdale/digital-garden/blob/main/index.md):

```md
Hi. I'm David Drysdale.

I'm a design strategist and researcher based in London, Ontario, Canada.

It's my mission to help people, teams, and organizations think more strategically, make better decisions, and achieve better results.
```

I was curious how that was magically converting to a beautiful `<h2>` tag, until I realized the homepage says "outcomes" not "results". 

```html
<h2>Hi. I'm David Drysdale.</h2>
<p>I’m a design strategist and researcher based in London, Ontario, Canada.</p>
<p>It’s my mission to help people, teams, and organizations think more strategically, make better decisions, and achieve better outcomes.</p>
```
&mdash; via https://github.com/djdrysdale/digital-garden/blob/50062eb4b362c381cb01c75952e7e9bcacd622d1/layouts/index.html#L45-L47

So even though you have content in your /index.md, your layout template hardcodes the same info (with formatting, and a different word) and ignores the original /index.md content. If you wanted, you could fix that by removing the duplicated bio from the layout and using `{{ content }}` in your /layouts/index.html file:

```liquid
                 <article>
                    {{ content }}
                 </article>
```

And then doing the formatting in Markdown in your /index.md file:

```md
## Hi. I'm David Drysdale.

I'm a design strategist and researcher based in London, Ontario, Canada.

It's my mission to help people, teams, and organizations think more strategically, make better decisions, and achieve better results.
```

Or just delete the unused content in /index.md and just have the frontmatter.

**Reply by djdrysdale** [reply]
Oh wow, thanks so much for this! I'm not the most comfortable with Eleventy (or anything to do with code, etc, for that matter) so I'm still processing everything. 

I tried to implement the changes, starting with just trying to get the two different kinds of content on the page, but I think I've made it worse. I am not sure if I understood the first step. That is my repo, but it was a bit outdated--I just pushed the current version of my site to GitHub.

**Reply by djdrysdale** [reply]
Oh wait--I got it to work! Going through the git diff showed me what I had done wrong. Again, much appreciated--you solved my problem and I can see I have a lot of polishing to do. :)

**Comment by pdehaan** [comment]
Also, are you using Nunjucks or LiquidJS or some other templating engine?
Solutions might be subtly different depending on what the respective engine's built-in filters are.

**Reply by djdrysdale** [reply]
I think it is using nunjucks. I started with eleventy-garden and then added a blog to it. Not sure if that helps.

**Reply by pdehaan** [reply]
Actually, I do believe that [liquid](https://liquidjs.com/tags/include.html#Passing-Variables) is the default template engine for .md and .html files
https://www.11ty.dev/docs/config/#default-template-engine-for-markdown-files
